2.14 release notes (#9406)
grafana/mimir: docs/sources/mimir/release-notes/v2.14.md

---
title: "Grafana Mimir version 2.14 release notes"
menuTitle: "V2.14 release notes"
description: "Release notes for Grafana Mimir version 2.14"
weight: 1100
---

# Grafana Mimir version 2.14 release notes

<!-- vale Grafana.GoogleWill = NO -->
<!-- vale Grafana.Timeless = NO -->
<!-- Release notes are often future focused -->

Grafana Labs is excited to announce version 2.14 of Grafana Mimir.

The highlights that follow include the top features, enhancements, and bug fixes in this release.
For the complete list of changes, refer to the [CHANGELOG](https://github.com/grafana/mimir/blob/main/CHANGELOG.md).

## Features and enhancements

> The minimal supported version of Go is updated to 1.22.

Reviewer: Is this relevant to users? They just use the binary we build, right?

Reviewer 2: As Charles says, this is not relevant. Furthermore mimir binaries are built with Go 1.23.1.

> The streaming of the chunks from store-gateways to queriers is enabled by default.

Author: Added additional note about the feature

Alertmanager adds the `-alertmanager.log-parsing-label-matchers` configuration option to control logging when parsing label matchers.
This flag is intended to be used with `-alertmanager.utf8-strict-mode-enabled` to validate UTF-8 strict mode is working as intended.

> The default value for the configuration option is `false`.

Author: I'm removing it from the notes.

> Alertmanager adds the `-alertmanager.utf8-migration-logging-enabled` configuration option, that enables logging of tenant configurations,

Reviewer 3: The alertmanager flags, compactor alert, and distributor changes don't seem like they're worth calling out in release notes.

> incompatible with UTF-8 strict mode. The default value for the configuration option is `false`.

Author: Reading it now, I have no idea, why I decided to add it to the notes ✂️

Compactor adds a new `cortex_compactor_disk_out_of_space_errors_total` counter metric, that tracks how many times a compaction failed
due to the compactor being out of disk. A new related alert `MimirCompactorHasRunOutOfDiskSpace` and it's runbook are added.

Incoming OTLP requests were previously size-limited with the distributor's `-distributor.max-recv-msg-size` configuration.
Distributor has a new `-distributor.max-otlp-request-size` configuration for limiting OTLP requests. The default value is set to limit request to 100 MiB.

> Distributor now replies with `Retry-After` header on retryable errors. The feature is controlled by the `-distributor.retry-after-header.enabled` configuration,

Reviewer 4: i think we now just enable it by default. Maybe we can mention what this means for users? It's that recovering from a Mimir outage should be easier now as the prometheus instances should respect Retry-After. ideally we can also mention how they should configure prometheus to respect retry-after headers. I'm not sure what the impact is for OTEL users

> Distributor now replies with `Retry-After` header on retryable errors. The feature is controlled by the `-distributor.retry-after-header.enabled` configuration,

Author: Added additional note about the feature.

that is enabled by default.

> 

Reviewer 2: One new feature that we could mention is support for marking ingesters as read-only, that is used in updated downscaling procedure (PR to be merged: https://github.com/grafana/mimir/pull/9208)

## Important changes

In Grafana Mimir 2.14 the following behavior has changed:

When executing a remote read request, the querier now honors the start/end time range, specified in the read hints.

> To prevent queue starvation with querier-worker queue prioritization algorithm, the querier now sets the minimum `-querier.max-concurrent` to four.

Reviewer 3: This seems pretty obscure

Values below the allowed minimum are ignored.

Distributor configuration's `-distributor.retry-after-header.max-backoff-exponent` and `-distributor.retry-after-header.base-seconds` are replaced
with `-distributor.retry-after-header.min-backoff` and `-distributor.retry-after-header.max-backoff` for easier configuration.

> The circuit breakers in ingester are improved to not open in case of per-instance limit errors. Since now on the circuit breakers

Reviewer 3: This is an implementation detail and probably doesn't need to be in release notes.

> The circuit breakers in ingester are improved to not open in case of per-instance limit errors. Since now on the circuit breakers

Author: I've removed mentioning of this one, as suggested above.

trigger the opening only if push and pull requests exceed the configured duration.

The default inactivity timeout of active series in ingesters, controlled by the `-ingester.active-series-metrics-idle-timeout` configuration
is increased from `10m` to `20m`.

> Query-frontend now returns the "413 Request Entity Too Large" error when a response shard from the `/active_series` request exceeds the limit.

Reviewer 3: This doesn't really need to be called out.

Following featues of store-gateway are changed: `-blocks-storage.bucket-store.max-concurrent-queue-timeout` is set to 5 seconds;
`-blocks-storage.bucket-store.index-header.lazy-loading-concurrency-queue-timeout` is set to 5 seconds;
`-blocks-storage.bucket-store.max-concurrent` is set to 200;

> 

Author: I'm not sure we should do it here. Nothing in the code related to redis configuration parameters are updated to be deprecated.

Number of previously deprecated configuration options were removed in the release:

- the `-ingester.return-only-grpc-errors` option in ingester
- the `-ingester.client.circuit-breaker.*` options in ingester
- the `-ingester.limit-inflight-requests-using-grpc-method-limiter` option in ingester
- the `-ingester.client.report-grpc-codes-in-instrumentation-label-enabled` option in distributor and ruler
- the `-distributor.limit-inflight-requests-using-grpc-method-limiter` option in distributor
- the `-distributor.enable-otlp-metadata-storage` option in distributor
- the `-ruler.drain-notification-queue-on-shutdown` option in ruler
- the `-querier.max-query-into-future` option in querier
- the `-querier.prefer-streaming-chunks-from-store-gateways` option in querier and store-gateway

> - the `-query-scheduler.use-multi-algorithm-query-queue` option in querier-scheduler

Reviewer 5: Do we make note of _all_ configuration options removed, or just the stable/non-experimental ones? This one was experimental.

- the YAML configuration `frontend.align_queries_with_step` in query-frontend

## Experimental features

Grafana Mimir 2.14 includes new features that are considered experimental and disabled by default.
Use them with caution and report any issues you encounter:

> A new experimental Kafka-based ingest storage architecture is added, that decouples the write and read path through a Kafka-compatible backend.

Reviewer 3: I'm not sure we want to mention this in release notes since it doesn't have any documentation yet or a migration path.

> A new experimental Kafka-based ingest storage architecture is added, that decouples the write and read path through a Kafka-compatible backend.

Reviewer 6: ```suggestion
There's an experimental Kafka-based ingest storage architecture that decouples the write and read paths through a Kafka-compatible backend.
```

> When enabled, distributors write incoming write requests to the Kafka-compatible backend, and the ingesters asynchronously replay ingested data from Kafka.

Reviewer 6: ```suggestion
When enabled, distributors write incoming write requests to the Kafka-compatible backend and the ingesters asynchronously replay ingested data from Kafka.
```

> New related configuration options that enable and configure the new architecture in Mimir are: `-ingest-storage.enabled`, `-ingest-storage.kafka.*`,

Reviewer 6: ```suggestion
The following configuration options enable and configure the new architecture in Mimir: `-ingest-storage.enabled`, `-ingest-storage.kafka.*`,
```

`-ingest-storage.ingestion-partition-tenant-shard-size`, `-ingest-storage.read-consistency`, `-ingest-storage.migration.distributor-send-to-ingesters-enabled`,

> `-ingester.partition-ring.*`.

Reviewer 6: I think we could make this am unordered list instead of comma-seperated values for greater readability.

Author: I've removed this one from the notes. The feature is still in the making.

> Querier has an experimental streaming PromQL engine. The configuration option `-querier.query-engine=mimir` enables the engine.

Reviewer: I wouldn't mention this yet, I wouldn't want to encourage people to start using it just yet.

> Querier has an experimental streaming PromQL engine. The configuration option `-querier.query-engine=mimir` enables the engine.

Reviewer 3: This should probably be the highlight of the release and deserves a longer description

> Querier has an experimental streaming PromQL engine. The configuration option `-querier.query-engine=mimir` enables the engine.

Reviewer 6: ```suggestion
The querier has an experimental streaming PromQL engine enabled by the`-querier.query-engine=mimir` configuration option.
```

> Ingester added experimental `-ingester.ignore-ooo-exemplars` configuration. When set, the out of order exemplars are no longer reported

Reviewer 6: ```suggestion
The ingester added an experimental `-ingester.ignore-ooo-exemplars` configuration. When set, out-of-order exemplars are no longer reported
```

to the remote write client.

> Querier now supports the experimental `limitk()` and `limit_ratio()` PromQL functions. The feature is disabled by default,

Reviewer 6: ```suggestion
The querier supports the experimental `limitk()` and `limit_ratio()` PromQL functions. This feature is disabled by default,
```

> but can be enabled with the `-querier.promql-experimental-functions-enabled=true` setting in the query-frontend and querier.

Reviewer 6: ```suggestion
but you can enable it with the `-querier.promql-experimental-functions-enabled=true` setting in the query-frontend and the querier.
```

Querier supports for the experimental `X-Mimir-Chunk-Info-Logger` header, that triggers logging information about TSDB chunks loaded from ingesters and store-gateways in the querier.

> The header should contain the comma separated list of labels, for which their value will be included in the logs.

Reviewer: I'm not sure this is worth mentioning, this is a feature intended to aid with debugging incorrect query results.

> The header should contain the comma separated list of labels, for which their value will be included in the logs.

Reviewer 6: ```suggestion
The header contains a comma-separated list of labels with values included in the logs.
```

Ruler includes the experimental `-ruler.rule-evaluation-write-enabled` configuration, that disables writing the result of rule evaluation to ingesters.

> This feature is meant to be used for testing.

Reviewer 6: ```suggestion
This feature is for testing purposes.
```

## Bug fixes

> - Ruler: add support for draining any outstanding alert notifications before shutting down. This can be enabled with the `-ruler.drain-notification-queue-on-shutdown=true` CLI flag.

Reviewer 6: ```suggestion
- Ruler: add support for draining any outstanding alert notifications before shutting down.  Enable this setting with the `-ruler.drain-notification-queue-on-shutdown=true` CLI flag.
```

- Query-frontend: "query stats" log now includes the actual `status_code` when the request fails due to an error occurring in the query-frontend itself.
- Store-gateway: fixed a case where, on a quick subsequent restart, the previous lazy-loaded index header snapshot was overwritten by a partially loaded one.
- Ingester: fixed timestamp reported in the "the sample has been rejected because its timestamp is too old" error when the write request contains only histograms.
- Store-gateway: store sparse index headers atomically to disk.
- Query scheduler: fix a panic in request queueing.
- Querier: fix issue where "context canceled" is logged for trace spans for requests to store-gateways that return no series when chunks streaming is enabled.
- Alertmanager: Fix per-tenant silence limits not reloaded during runtime.
- Alertmanager: Fixes a number of bugs in silences which could cause an existing silence to be deleted/expired when updating the silence failed. This could happen when the replacing silence was invalid or exceeded limits.
- Alertmanager: Fix help message for utf-8-strict-mode.
- Query-frontend: Ensure that internal errors result in an HTTP 500 response code instead of 422.

> - Configuration: Multi line envs variables are flatten during injection to be compatible with YAML syntax.

Reviewer 6: ```suggestion
- Configuration: Multi-line environment variables are flattened during injection to be compatible with YAML syntax.
```

- Querier: fix issue where queries can return incorrect results if a single store-gateway returns overlapping chunks for a series.
- HA Tracker: store correct timestamp for last received request from elected replica.
- Querier: do not return `grpc: the client connection is closing` errors as HTTP `499`.
- Compactor: fix a race condition between different compactor replicas that may cause a deleted block to be still referenced as non-deleted in the bucket index.
- Querier: fix issue where some native histogram-related warnings were not emitted when `rate()` was used over native histograms.
- Ruler: map invalid org-id errors to 400 status code.
- Querier: Fix invalid query results when multiple chunks are being merged.
- Query-frontend: return annotations generated during evaluation of sharded queries.
- Querier: Support optional start and end times on `/prometheus/api/v1/labels`, `/prometheus/api/v1/label/<label>/values`, and `/prometheus/api/v1/series` when `max_query_into_future: 0`.

> - Alertmanager: Fix config validation gap around unreferenced templates.

Reviewer 6: ```suggestion
- Alertmanager: Fix configuration validation gap around unreferenced templates.
```

> - Alertmanager: Fix goroutine leak when stored config fails to apply and there is no existing tenant alertmanager.

Reviewer 6: ```suggestion
- Alertmanager: Fix goroutine leak when stored configuration fails to apply and there is no existing tenant alertmanager.
```
